### PR TITLE
[swift-3-indexing-model] porting Foundation to the new indexing model

### PR DESCRIPTION
**Pull request by gribozavr**
This PR contains adjustments for SE-0065 "A New Model for Collections and Indices".

### CODE REVIEW

File: Foundation/NSData.swift [hunk 1]
```diff
@@ -539,7 +539,14 @@ extension NSData {
...
     }
-    private static func searchSubSequence<T : Collection,T2 : Sequence where T.Iterator.Element : Equatable, T.Iterator.Element == T2.Iterator.Element, T.SubSequence.Iterator.Element == T.Iterator.Element>(_ subSequence : T2, inSequence seq: T,anchored : Bool) -> T.Index? {
+    private static func searchSubSequence<
+        T : Collection,T2 : Sequence
+        where
+        T.Iterator.Element : Equatable,
+        T.Iterator.Element == T2.Iterator.Element,
+        T.SubSequence.Iterator.Element == T.Iterator.Element,
+        T.Indices.Iterator.Element == T.Index
+    >(_ subSequence : T2, inSequence seq: T,anchored : Bool) -> T.Index? {
```
Comment (parkera) [inline Foundation/NSData.swift]: Let's leave this patch to fundamentals and not changing of the coding style.

**Comment by parkera** [comment]
Looks ok to me except for the minor comment about the code style change. Merge when you're ready.

File: Foundation/NSData.swift (hunk 1 above)
Comment (gribozavr) [inline Foundation/NSData.swift]: Fixed, pushed an update.

**Comment by gribozavr** [comment]
Thanks @parkera!

**Comment by mbvreddy** [comment]
foundation fails to build due to below compilation error.

```
Foundation/String.swift:86:28: error: 'Indices' is not a member type of 'String.UTF16View'
extension String.UTF16View.Indices : RandomAccessCollection {}
          ~~~~~~~~~~~~~~~~ ^
ninja: build stopped: subcommand failed.
```

**Comment by gribozavr** [comment]
@mbvreddy How are you building?  What git hashes do you have checked out?  The buildbots were happy since I merged it: https://ci.swift.org/job/oss-swift-incremental-RA-linux-ubuntu-14_04/4510/

**Comment by mbvreddy** [comment]
@gribozavr 
I was building only foundation using "ninja". I now tried to build complete foundation using "./utils/build_script --foundation --xctest" and it is built successfully.
Thank you.

**Pull request merged**